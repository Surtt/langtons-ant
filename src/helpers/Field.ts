type Cell = false | true;
type Line = Cell[];
export type Field = Line[];

export const generateField = (size: number, state: Cell): Field =>
  new Array<boolean[]>(size)
    .fill(new Array<boolean>(size))
    .map(() => new Array<boolean>(size).fill(state));

// export const toggleCell = (cell: Cell, coords: Coords): Cell => { };
