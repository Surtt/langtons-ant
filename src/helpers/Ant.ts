import { Cell } from '@/helpers/Field';
export type Coords = [number, number];

type RandomGenerator<R = number> = (
  min: number,
  max: number,
  rotation?: number
) => R;

export type Ant = {
  xAnt: number;
  yAnt: number;
  rotation: number;
};

export const generateNumber: RandomGenerator = (min, max) => {
  return Math.floor(Math.random() * (max - min)) + min;
};

export const generateAntPosition: RandomGenerator<Ant> = (
  min: number,
  max: number,
  rotation = 0
) => {
  const xAnt = generateNumber(min, max);
  const yAnt = generateNumber(min, max);
  return {
    xAnt,
    yAnt,
    rotation,
  };
};

export const changeAntState = (
  rotation: number,
  cell: Cell
): {
  xAnt: number;
  yAnt: number;
} => {
  const value = { xAnt: 0, yAnt: 0 };
  switch (rotation) {
    case 90:
      value.xAnt++;
      break;
    case 180:
      value.yAnt--;
      break;
    case 270:
      value.xAnt--;
      break;
    default:
      value.yAnt++;
      break;
  }
  if (cell) {
    value.xAnt = -value.xAnt;
    value.yAnt = -value.yAnt;
  }
  return value;
};

export const newRotation = (rotation: number, cell: Cell): number => {
  let newRotation = rotation + 90;
  if (cell) {
    newRotation += 180;
  }
  if (newRotation >= 360) {
    newRotation -= 360;
  }
  return newRotation;
};
