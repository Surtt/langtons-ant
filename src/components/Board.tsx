import React, { FC } from 'react';
import { Line } from './Line';
import { generateField } from '../helpers/Field';
import { generateAntPosition } from '../helpers/Ant';

export const Board: FC = () => {
  const basicSize = 15;
  const defaultCellState = false;
  const cells = generateField(basicSize, defaultCellState);
  const ant = generateAntPosition(0, basicSize);
  return (
    <div className="stretch">
      <table className="board">
        <tbody>
          {cells.map((line, index) => (
            <Line key={`line_${index}`} line={line} index={index} ant={ant} />
          ))}
        </tbody>
      </table>
    </div>
  );
};
