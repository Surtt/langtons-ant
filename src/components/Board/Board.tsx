import React, { FC } from 'react';
import styled from '@emotion/styled';
import { Line } from '../Line/Line';
import { Field } from '@/helpers/Field';
import { Ant } from '@/helpers/Ant';

export interface BoardProps {
  /**
   * Field data
   */
  cells: Field;
  /**
   * Ant position
   */
  ant: Ant;
}

export const Board: FC<BoardProps> = ({ cells, ant }) => {
  return (
    <BoardTable>
      <tbody>
        {cells.map((line, index) => (
          <Line key={`line_${index}`} line={line} index={index} ant={ant} />
        ))}
      </tbody>
    </BoardTable>
  );
};

export const BoardTable = styled.table`
  width: 100%;
  height: 100%;
  border-collapse: collapse;
  border: 1px solid #337ab7;
  margin-bottom: 1em;
`;
