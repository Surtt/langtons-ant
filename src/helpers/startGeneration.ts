import { changeAntState, newRotation, Ant } from './Ant';
import { Field, toggleCellOnTheField } from './Field';
export interface GeneratingValues {
  field: Field;
  ant: Ant;
  count: number;
}

const isFieldBorder = (field: Field, ant: Ant) => {
  const borderY = field.length - 1;
  const borderX = field[0].length - 1;
  return (
    ant.yAnt === 0 ||
    ant.yAnt === borderY ||
    ant.xAnt === 0 ||
    ant.xAnt === borderX
  );
};

export const startGeneration = (
  field: Field,
  ant: Ant,
  count: number
): GeneratingValues => {
  const cnt = count + 1;
  const { xAnt, yAnt, rotation } = ant;

  const movement = changeAntState(rotation, field[xAnt][yAnt]);
  const updateRotation = newRotation(rotation, field[xAnt][yAnt]);
  field = toggleCellOnTheField(field, [xAnt, yAnt]);
  movement.xAnt += xAnt;
  movement.yAnt += yAnt;
  const newAnt = {
    xAnt: movement.xAnt,
    yAnt: movement.yAnt,
    rotation: updateRotation,
  };

  if (isFieldBorder(field, ant)) {
    const length = field.length;
    field.unshift(new Array<boolean>(length).fill(false));
    field.push(new Array<boolean>(length).fill(false));
    field.forEach((item) => {
      item.unshift(false);
      item.push(false);
    });
    if (newAnt) {
      newAnt.xAnt++;
      newAnt.yAnt++;
    }
  }

  return {
    field: [...field],
    ant: newAnt,
    count: cnt,
  };
};
