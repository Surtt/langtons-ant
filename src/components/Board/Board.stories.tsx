import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Board, BoardProps } from './Board';
import { StretchWrap } from '@/App';

export default {
  title: 'Board/Board',
  component: Board,
  decorators: [
    (Story) => <StretchWrap>
      <Story />
    </StretchWrap>
  ]
} as Meta;

const Template: Story<BoardProps> = (args) => <Board {...args} />;

export const BoardWithAnt = Template.bind({});
BoardWithAnt.args = {
  cells: [
    [false, false, false, false, false, false, false, false, false, false],
    [false, false, false, false, false, false, false, false, false, false],
    [false, false, false, false, false, false, false, false, false, false],
    [false, false, false, false, false, false, false, false, false, false],
    [false, false, false, false, false, false, false, false, false, false],
    [false, false, false, false, false, false, false, false, false, false],
    [false, false, false, false, false, false, false, false, false, false],
    [false, false, false, false, false, false, false, false, false, false],
    [false, false, false, false, false, false, false, false, false, false],
    [false, false, false, false, false, false, false, false, false, false],
  ],
  ant: { xAnt: 4, yAnt: 4, rotation: 0 }
};
