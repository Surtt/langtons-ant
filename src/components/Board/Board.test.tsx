import React from 'react';
import { render, screen } from '@testing-library/react';
import { Board } from './Board';
import { Field } from '@/helpers/Field';
import { Ant } from '@/helpers/Ant';

const cells: Field = [
  [false, false, false, false, false],
  [false, false, false, false, false],
  [false, false, false, false, false],
  [false, false, false, false, false],
  [false, false, false, false, false],
];

const ant: Ant = {
  xAnt: 0,
  yAnt: 0,
  rotation: 0,
};

it('Renders Board', () => {
  const { asFragment } = render(<Board cells={cells} ant={ant} />);
  expect(asFragment()).toMatchSnapshot();
});

it('Renders Board role', () => {
  render(<Board cells={cells} ant={ant} />);
  expect(screen.getByRole('grid')).toBeInTheDocument();
});
