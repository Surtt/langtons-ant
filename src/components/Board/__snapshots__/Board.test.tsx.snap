// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Board 1`] = `
<DocumentFragment>
  <table
    class="css-1abeprg"
    role="grid"
  >
    <tbody>
      <tr
        role="line"
      >
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class="css-fzp5ig"
              rotate="0"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
      </tr>
      <tr
        role="line"
      >
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
      </tr>
      <tr
        role="line"
      >
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
      </tr>
      <tr
        role="line"
      >
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
      </tr>
      <tr
        role="line"
      >
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
        <td
          class="css-1dvq4hu"
          role="cell"
        >
          <div
            class="css-1h0vbn9"
          >
            <div
              class=" css-bdbrua"
            />
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
